### DISCUSSION TITLE
Best way to know if `CrudRepository.deleteById` deleted something

### DISCUSSION THREAD

**Discussion by KarlHakanNordgren** [micronaut-projects/micronaut-data#1535] [Q&A] [answered]
Hi All! If there is no record with `id0` as its id then `crudRepository.deleteById(id0)` will complete silently having done nothing. What is the best way to only delete if a record exists? Is the below performant?

```
if (crudRepository.existsById(id)) {
    crudRepository.deleteById(id)
} else { throw SomeException() }
```

**Comment by dstepanov** [comment, marked as answer]
You can create an alternative method ```Long removeById(Id)``` and it will return the number of records deleted.

**Reply by KarlHakanNordgren** [reply]
Pure magic! Could I ask where this method is documented?

**Reply by dstepanov** [reply]
https://micronaut-projects.github.io/micronaut-data/latest/guide/#deletes

**Reply by graemerocher** [reply]
one improvement to the docs would be to mention you can return a number which we don't see to mention

**Reply by dstepanov** [reply]
We should have some table with all possible method prefixes and return types

**Reply by KarlHakanNordgren** [reply]
Wait according to that I can do `Long deleteById(Id)`. Is that correct?

**Reply by dstepanov** [reply]
You can if you don’t use one of the existing repositories that defines no return methods

**Reply by KarlHakanNordgren** [reply]
Ps. Thanks for the above. Worked very nicely!

**Comment by theMyth721** [comment]
Thank you but I am facing the same issue:
1) I can't find any such removeById method in any of the repositories
2) The deleteById just silently succeeds regardless of weather the record exists or not,
For this i think I need to create a custom solution or it would take two queries to sort this unfortunately which is not an efficient solution.
I cannot believe such a mature framework did not think of something as obvious as this, I cannot see the reason behind it

**Reply by dstepanov** [reply]
You can create a new interface and create a method there.

**Reply by theMyth721** [reply]
Yes, typically I will be creating my own implementations with my own stored procedures, but without that, the only way is with 2 queries to the database.
Thank you so much for getting back

**Reply by dstepanov** [reply]
There is a solution mentioned in this thread, you can add a new method to your repository `Long removeById(Id)` that one will return the number of rows affected, so you will know if the record is deleted. You can also define:

```
default void deleteByIdOrFail(Long id) {
   if (removeById(id) == 0) {
       throw ...
   }
}
```
